README
= Columnize - Module to show an Array formatted as a String aligned in columns. 

== Summary

In showing a long lists, sometimes one would prefer to see the value
arranged aligned in columns. Some examples include listing methods of
an object, listing debugger commands, or showing a numeric array with data
aligned.

=== Setup

   use Array::Columnize;
   print columnize(['a','b','c','d'], {displaywidth=>4});

produces:

    1  3
    2  4

=== With numeric data

    $data_ref = [80..120];
    print columnize($data, {ljust = 0}) ;

produces:

    80  83  86  89  92  95   98  101  104  107  110  113  116  119
    81  84  87  90  93  96   99  102  105  108  111  114  117  120
    82  85  88  91  94  97  100  103  106  109  112  115  118

while:

   print columnize($data_ref, {ljust = 0, arrange_verticle = 0}) ;

produces:

   80  83  86  89  92  95   98  101  104  107  110  113  116  119
   81  84  87  90  93  96   99  102  105  108  111  114  117  120
   82  85  88  91  94  97  100  103  106  109  112  115  118

=== With String data

  $aref = qw(bibrons golden madascar leopard mourning suras tokay);
  print columnize $g, :displaywidth => 15
  $ print columnize($aref, {displaywidth => 15});
  bibrons   suras
  golden    tokay
  madascar  
  leapard
  mourning

  $ puts columnize $aref, {displaywidth => 18, colsep => ' | '};
  bibrons  | suras
  golden   | tokay
  madascar
  leopard 
  mourning

== Credits

  This is adapted from my Ruby gem of the same name.

== Other stuff

Author::   Rocky Bernstein <rockyb@rubyforge.net>
License::  Copyright (c) 2011 Rocky Bernstein
           Released under the GNU GPL 2 license

== Warranty

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


Changes
Revision history for Array::Columnize

1.04 2013-06-12
    - botched release 1.03

1.03 2013-06-12
    - Add linesuffix, and colfmt options
    - document all options
    - Work around rlib limitation.

1.02 2012-08-19
    - Rerelease just for some Meta and Doc changes

1.01 2012-08-13
    - Add option arrange_array to format as an array.
    - Use rlib more pervasivly and use it
    - incompatible change: always end string with "\n"

1.00 2012-08-12
    - Packaging changes. Version number conforms to
      http://perldoc.perl.org/perlmodstyle.html#Version-numbering
    - Change Meta data to point to github repository and issue tracker

0.4.2 2012-01-02
    - Re-release to reduce dependencies

0.4.1 2011-12-03
    - Try to silence CPANTS

0.4.0 2011-11-24
    - Handle situations where text is longer than display width.
    - More strict on tests and as per CPANTS reports

0.3.9 2011-11-12
    - rerelease with 'dev' in version number.
      it was messing up some Module::Builds
    - Update GNUmakefile
    - Build.PL back to more convenient form (for worse kwalitee).

0.3.8 2011-10-29
    - Made the Build.PL non-executable (for better kwalitee:
        - http://cpants.perl.org/kwalitee.html#buildtool_not_executable
    - Add an environment variable - PERL_MB_NOSIGN to disable signing a package
    (which interferes with the flow of doing ./Build disttest ).
    - Renamed Makefile to Makefile.devel so doing perl Makefile.PL won't
    interfere with it.
    - Correct some typos in the README.md and added a TODO there (SHLOMIF).
    - Add some POD to lib/Array/Columnize.pm (SHLOMIF).

0.3.7 2011-10-28
I said, no feature "say"!

0.3.6 2011-10-27
Remove feature "say" so it works on Perl pre 5.010.

0.3.5 2011-09-05
Initial revision


Build.PL
#!/usr/bin/env perl
#  Copyright (C) 2011-2013 Rocky Bernstein <rocky@cpan.org>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use Module::Build;
use Config;
use File::Basename;
use File::Spec;
my $lib = File::Spec->catfile(dirname(__FILE__), 'lib');
unshift @INC, $lib;
require Array::Columnize;

my $release_status = ($Array::Columnize::VERSION =~ /_/) ?
    'testing' : 'stable';

my $builder = Module::Build->new(
    module_name         => 'Array::Columnize',
    add_to_cleanup      => [ 'Array-Columnize-*', 'tmp*'  ],
    create_makefile_pl  => 'passthrough',
    dist_abstract       =>
    'Format an Array as an Array of String aligned in columns.',
    dist_author         => 'Rocky Bernstein <rocky@cpan.org>',
    dist_version_from   => 'lib/Array/Columnize.pm',
    license             => 'gpl',
    needs_compiler      => 0,
    release_status      => $release_status,

    meta_merge => {
	resources => {
	    bugtracker  => 'https://github.com/rocky/Perl-Array-Columnize/issues',
	    repository  => 'http://github.com/rocky/Perl-Array-Columnize'
	}
    },


    configure_requires => {
	'Module::Build'           => 0,
	'perl'                    => '5.006001',
        'rlib'                    => '0.02',
	'File::Spec'              => 0,
	'File::Basename'          => 0,
    },
    build_requires => {
        'Test::More'              => 0.81,
        'rlib'                    => '0.02',
    },
    requires => {
        'version'                 => 0,
        'rlib'                    => '0.02',
    },
    sign                => ($ENV{PERL_MB_NOSIGN} ? 0 : 1),
    );

$builder->create_build_script();


lib/Array/Columnize/options.pm
=head1 DESCRIPTION

# Array::Columnize options processing

=cut
package Array::Columnize;

use vars qw($DEFAULT_OPTS);

# Default values for columize options.
$DEFAULT_OPTS = {
    arrange_array => 0,
    arrange_vertical => 1,
    array_prefix => '',
    array_suffix => '',
    colsep => '  ',
    displaywidth => $ENV{'COLUMNS'} || 80,
    lineprefix => '',
    ljust => 'auto',
    term_adjust => 0
};

# Merge in default configuration options into the passed hash reference.
# Values already set in the hash are untouched.
sub merge_config(%) {
    my $config = shift;
    while (($field, $default_value) = each %$DEFAULT_OPTS) {
	$config->{$field} = $default_value unless defined $config->{$field};
    };
}

if (__FILE__ eq $0 ) {
    my %config;
    merge_config \%config;
    require Data::Dumper;
    print Data::Dumper::Dumper(\%config), "\n";

    my $config = {
	arrange_array => 0,
	term_adjust   => 1,
	lineprefix    => '...',
	displaywidth  => 10,
	bogus         => 'yep'
    };
    print Data::Dumper::Dumper($config), "\n";
    merge_config $config;
    print Data::Dumper::Dumper($config), "\n";
}

1;


lib/Array/Columnize/columnize.pm
#!/usr/bin/env perl
# See doc in Array::Columnize
BEGIN {
    no strict;
    @OLD_INC = @INC;
}
use rlib '../..';
use Array::Columnize::options;
BEGIN {
    no strict;
    @INC = @OLD_INC;
}

use Array::Columnize::options;
package Array::Columnize;
use strict;
use warnings;
use POSIX;

=pod

=head1 Subroutines

=head2 cell_size

Return the length of String I<cell>. If Boolean I<term_adjust> is true,
ignore terminal sequences in I<cell>.

=cut

sub cell_size($$) {
    my ($cell, $term_adjust) = @_;
    $cell =~ s/\e\[.*?m//g if $term_adjust;
    return length($cell);
}

=head2 columnize

Return a list of strings with embedded newlines (\n) as a compact
set of columns arranged horizontally or vertically.

For example, for a line width of 4 characters (arranged vertically):

     ['1', '2,', '3', '4'] => '1  3\n2  4\n'

or arranged horizontally:

     ['1', '2,', '3', '4'] => '1  2\n3  4\n'

Each column is only as wide possible, no larger than
C<$opts->{displaywidth}>.  If I<aref>is not an array reference, the
empty string, '', is returned. By default, columns are separated by
two spaces - one was not legible enough. Set C<$opts->{colsep}> to
adjust the string separate columns. If C<$opts->{arrange_vertical} is
set false, consecutive items will go across, left to right, top to
bottom.

=cut

sub columnize($;$) {
    my($aref, $opts) = @_;
    my @l = @$aref;

    # Some degenerate cases
    # FIXME test for arrayness
    # return '' if  $aref is not an array
    return "<empty>\n" if 0 == scalar(@l);
    $opts = {} unless $opts;
    merge_config $opts;
    if ($opts->{arrange_array}) {
        $opts->{array_prefix} ||= '(';
	$opts->{lineprefix}   ||= ' ';
        $opts->{array_suffix} ||= ")";
        $opts->{colsep}         = ', ';
        $opts->{arrange_vertical} = 0;
    };
    if (1 == scalar @l) {
	my $ret = sprintf("%s%s%s\n",
			  $opts->{array_prefix},
			  $l[0],
			  $opts->{array_suffix});
	return $ret;
    }

    @l = map(sprintf($opts->{colfmt}, $_), @l) if $opts->{colfmt};

    my %opts = %$opts;

    my ($nrows, $ncols) = (0, 0);  # Make nrows, ncols have more global scope
    my @colwidths = ();     # Same for colwidths

    if ($opts{displaywidth} - length($opts{lineprefix}) < 4) {
	$opts{displaywidth} = length($opts{lineprefix}) + 4;
    } else {
	$opts{displaywidth} -= length($opts{lineprefix})
    }
    if ($opts{arrange_vertical}) {
	my $array_index = sub ($$$) {
	    my ($num_rows, $row, $col) = @_;
	    ($num_rows * $col) + $row
	};
	# Try every row count from 1 upwards
	for (my $_nrows=1; $_nrows < scalar @l; $_nrows++) {
	    $nrows = $_nrows;
	    $ncols = POSIX::ceil((scalar(@l)) / $nrows);
	    @colwidths = ();
	    my $totwidth = -length($opts{colsep});

	    for (my $col=0; $col < $ncols; $col++) {
		# get max column width for this column
		my $colwidth = 0;
		for (my $row=0; $row < $nrows; $row++) {
		    my $i = $array_index->($nrows, $row, $col);
		    last if ($i >= scalar(@l));
		    my $try_width = cell_size($l[$i], $opts{term_adjust});
		    $colwidth =  $try_width if $try_width > $colwidth;
		}
		push(@colwidths, $colwidth);
		$totwidth += $colwidth + length($opts{colsep});
		if ($totwidth > $opts{displaywidth}) {
		    $ncols = $col;
		    last;
		}
	    }
	    last if ($totwidth <= $opts{displaywidth});
	}
	$ncols = 1 if $ncols < 1;
	$nrows = scalar(@l) if $ncols == 1;

	# The smallest number of rows computed and the max widths for
	# each column has been obtained.  Now we just have to format
	# each of the rows.
	my @s = ();
	for (my $row=0; $row < $nrows; $row++) {
	    my @texts = ();
	    my $x;
	    for (my $col=0; $col < $ncols; $col++) {
		my $i = $array_index->($nrows, $row, $col);
		if ($i >= scalar(@l)) {
		    $x = '';
		} else {
		    $x = $l[$i];
		}
		push @texts, $x;
	    }
	    pop(@texts) while (scalar(@texts) > 0 && $texts[-1] eq '');
	    if (scalar(@texts) > 0) {
		for (my $col=0; $col < scalar(@texts); $col++) {
		    unless ($ncols == 1 && $opts{ljust}) {
			my $fmt = sprintf("%%%s$colwidths[$col]s",
					  ($opts{ljust} ? '-': ''));
			$texts[$col] = sprintf($fmt, $texts[$col]);
		    }
		}
		push(@s, sprintf("%s%s", $opts{lineprefix},
				 join($opts{colsep}, @texts)));
	    }
	}
	return join($opts{linesuffix}, @s) . "\n";
    } else {
    	my $array_index = sub ($$$) {
	    my ($num_rows, $row, $col) = @_;
	    $ncols * ($row-1) + $col;
	};
    	# Try every column count from size downwards.
    	my ($totwidth, $i, $rounded_size) = (0, 0, 0);
        for (my $_ncols=scalar(@l); $_ncols >= 1; $_ncols--) {
	    $ncols = $_ncols;
	    # Try every row count from 1 upwards
	    my $min_rows = POSIX::ceil((scalar(@l)+$ncols-1) / $ncols);
	    for (my $_nrows=$min_rows; $_nrows <= scalar(@l); $_nrows++) {
		$nrows = $_nrows;
		$rounded_size = $nrows * $ncols;
		@colwidths = ();
		$totwidth = -length($opts{colsep});
		my ($colwidth, $row) = (0,0);
		for (my $col=0; $col < $ncols; $col++) {
		    # get max column width for this column
		    for (my $_row=1; $_row <= $nrows; $_row++) {
		    	$row = $_row;
		    	$i = $array_index->($nrows, $row, $col);
		    	last if $i >= scalar(@l);
			my $try_size = cell_size($l[$i],
						 $opts{term_adjust});
			$colwidth = $try_size if $try_size > $colwidth;
		    }
		    push @colwidths, $colwidth;
		    $totwidth += $colwidth + length($opts{colsep});
		    last if ($totwidth > $opts{displaywidth});
		}
		if ($totwidth <= $opts{displaywidth}) {
		    # Found the right nrows and ncols
		    $nrows  = $row;
		    last;
		}
		elsif ($totwidth >= $opts{displaywidth}) {
		    # Need to reduce ncols
		    last;
		}
	    }
	    last if ($totwidth <= $opts{displaywidth} && $i >= $rounded_size-1);
	}
	$nrows = scalar(@l) if $ncols == 1;

	# The smallest number of rows computed and the
	# max widths for each column has been obtained.
	# Now we just have to format each of the
	# rows.
	my @s = ();
	my $prefix = $opts{array_prefix} = '' ?
	    $opts{lineprefix} : $opts{array_prefix};
	for (my $row=1; $row <= $nrows; $row++) {
	    my @texts = ();
	    my $x;
	    for (my $col=0; $col < $ncols; $col++) {
		my $i = $array_index->($nrows, $row, $col);
		if ($i >= scalar(@l)) {
		    last;
		} else {
		    $x = $l[$i];
		}
		push @texts, $x;
	    }
	    for (my $col=0; $col < scalar(@texts); $col++) {
		unless ($ncols == 1 && $opts{ljust}) {
		    my $fmt = sprintf("%%%s$colwidths[$col]s",
				      ($opts{ljust} ? '-': ''));
		    $texts[$col] = sprintf($fmt, $texts[$col]);
		}
	    }
	    push(@s, sprintf("%s%s", $prefix,
			     join($opts{colsep}, @texts))) if scalar(@texts);
	    $prefix = $opts->{lineprefix};
	}
	$s[-1] .= $opts->{array_suffix};
	return join($opts{linesuffix}, @s) . "\n";
    }
}


# Demo it
unless (caller) {

    my @ary = qw(bibrons golden madascar leopard mourning suras tokay);
    print columnize(\@ary, {displaywidth => 18});

    my $line = 'require [1;29m"[0m[1;37mirb[0m[1;29m"[0m';
    print cell_size($line, 1), "\n";
    print cell_size($line, 0), "\n";

    print columnize(['hi']), "\n";
    print columnize([]), "\n";

    for my $tuple ([4, 4], [4, 7], [100, 180]) {
	my @data = ($tuple->[0]..$tuple->[1]);
	print columnize(\@data, {colsep =>'  ', arrange_vertical=>0});
	print '------------------------';
	print columnize(\@data, {colsep =>'  ', arrange_vertical=>1});
	print '========================';
    }
    print columnize(["a", 2, "c"], {displaywidth => 10, colsep => ', '});
    print columnize(["oneitem"]);
    print columnize(["one", "two", "three"]);
    my @data = ("one",       "two",         "three",
		"for",       "five",        "six",
		"seven",     "eight",       "nine",
		"ten",       "eleven",      "twelve",
		"thirteen",  "fourteen",    "fifteen",
		"sixteen",   "seventeen",   "eightteen",
		"nineteen",  "twenty",      "twentyone",
		"twentytwo", "twentythree", "twentyfour",
		"twentyfive","twentysix",   "twentyseven");

    print columnize(\@data);
    @data = (1..30);
    print columnize(\@data,
		    {arrange_array => 1, ljust =>0, displaywidth => 70});
}

1;


lib/Array/Columnize.pm
# Format an Array as an Array of String aligned in columns.
#
# == Summary
# Format a list into a single string with embedded newlines.
# On printing the string the columns are aligned.
#
#  See documentation for Columnize.columnize below.
#
# == License
#
# Columnize is copyright (C) 2011-2013 Rocky Bernstein <rocky@cpan.org>
#
# All rights reserved.  You can redistribute and/or modify it under
# the same terms as Perl.
#
# Adapted from the routine of the same name in Ruby.

package Array::Columnize;
use strict;
use Exporter;
use warnings;
use lib '..';

use Array::Columnize::columnize;
use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);
@ISA = qw/ Exporter /;
@EXPORT = qw(columnize);

# Add or remove  _01 when we want testing.
use version; $VERSION = '1.04_01';

unless (caller) {
    # Demo code
    print "This is version: $Array::Columnize::VERSION\n";
    print columnize([1,2,3,4], {displaywidth=>4}), "\n";
    my $data_ref = [80..120];
    print columnize($data_ref, {ljust => 0}) ;
    print columnize($data_ref, {ljust => 0, arrange_vertical => 0}) ;
    my @ary = qw(bibrons golden madascar leopard mourning suras tokay);
    print columnize(\@ary, {displaywidth => 18});
}

'Just another Perl module';
__END__
=encoding utf8

=head1 NAME

Array::Columnize - arrange list data in columns.

=head1 SYNOPSIS

    use Array::Columnize;
    print columnize($array_ref, $optional_hash_or_hash_ref);

=head1 DESCRIPTION

In showing long lists, sometimes one would prefer to see the values
arranged and aligned in columns. Some examples include listing methods of
an object, listing debugger commands, or showing a numeric array with data
aligned.

=head2 OPTIONS

=over

=item displaywidth

the line display width used in calculating how to align columns

=item colfmt

What format specifier to use in sprintf to stringify list entries. The
default is none.

=item colsep

String to insert between columns. The default is two spaces, oe space
just wasn't enough.

the column separator

=item lineprefix

=item linesuffix

=item termadjust

=item arrange_array

=item ladjust

whether to left justify text instead of right justify. The default is true

=back

=head1 EXAMPLES

=head2 Simple data example

    print columnize(['a','b','c','d'], {displaywidth=>4});

produces:

    a  c
    b  d

=head2 With numeric data

    my $array_ref = [80..120];
    print columnize($array_ref, {ljust => 0}) ;

produces:

    80  83  86  89  92  95   98  101  104  107  110  113  116  119
    81  84  87  90  93  96   99  102  105  108  111  114  117  120
    82  85  88  91  94  97  100  103  106  109  112  115  118

while:

    print columnize($array_ref, {ljust => 0, arrange_vertical => 0}) ;

produces:

     80   81   82   83   84   85   86   87   88   89
     90   91   92   93   94   95   96   97   98   99
    100  101  102  103  104  105  106  107  108  109
    110  111  112  113  114  115  116  117  118  119
    120

And

    my $array_ref = [1..30];
    print columnize($array_ref,
		    {arrange_array => 1, ljust => 0, displaywidth => 70});

produces:

   ( 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30)

=head2 With String data

    @ary = qw(bibrons golden madascar leopard mourning suras tokay);
    print columnize(\@ary, {displaywidth => 18});

produces:

    bibrons   mourning
    golden    suras
    madascar  tokay
    leopard

    print columnize \@ary, {displaywidth => 18, colsep => ' | '};

produces:

    bibrons  | mourning
    golden   | suras
    madascar | tokay
    leopard

=head1 AUTHOR

Rocky Bernstein, C<< rocky@cpan.org >>

=head1 BUGS

Please report any bugs or feature requests through the web interface
at L<https://github.com/rocky/Perl-Array-Columnize/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Array::Columnize

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Array-Columnize>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Array-Columnize>

=item * Github request tracker

L<https://github.com/rocky/Perl-Array-Columnize/issues>

=item * Search CPAN

L<http://search.cpan.org/dist/Array-Columnize>

=back

=head1 COPYRIGHT & LICENSE

Copyright (c) 2011, 2012 Rocky Bernstein.

=head2 LICENSE

Same terms as Perl.

=cut
